src/tools/w3dviewer/graphicview.cpp

```diff
@@ -0,0 +1,1095 @@
+    m_mousePos = point;
+    m_rightButtonDown = TRUE;
+
+    if (m_leftButtonDown) {
+        SetCursor(LoadCursor(AfxGetResourceHandle(), MAKEINTRESOURCE(IDC_MOVE)));
+    } else {
+        SetCursor(LoadCursor(AfxGetResourceHandle(), MAKEINTRESOURCE(IDC_ZOOM)));
+    }
+
+    CWnd::Default();
+}
+
+void CGraphicView::OnGetMinMaxInfo(MINMAXINFO FAR *lpMMI)
+{
+    CWnd::Default();
+}
+
+static void CALLBACK TimerProc(UINT uTimerID, UINT uMsg, DWORD_PTR dwUser, DWORD_PTR dw1, DWORD_PTR dw2)
+{
+    if (dwUser != 0 && !GetProp((HWND)dwUser, "WaitingToProcess") && !GetProp((HWND)dwUser, "Inactive")) {
+        SetProp((HWND)dwUser, "WaitingToProcess", (HANDLE)1);
+        PostMessage((HWND)dwUser, GVM_RENDER, 0, 0);
+    }
+}
+
+BOOL CGraphicView::Create()
+{
+    if (Device < 0) {
+        return FALSE;
+    }
+
+    m_isInitialized = FALSE;
+    RECT r;
+    GetClientRect(&r);
+    int width = r.right - r.left;
+    int height = r.bottom - r.top;
+
+    if (!m_isWindowed) {
+        width = DeviceWidth;
+        height = DeviceHeight;
+    }
+
+    W3DErrorType error = W3D::Set_Render_Device(Device, width, height, BPP, m_isWindowed, true, false, true);
+    bool success = error == W3D_ERROR_OK;
+
+    if (error == W3D_ERROR_OK && m_camera == nullptr) {
+        m_camera = new CameraClass();
+
+        success = m_camera != nullptr;
+
+        if (m_camera != nullptr) {
+            Matrix3D m(true);
+            m[2].W = 35.0f;
+            m_camera->Set_Transform(m);
+        }
+    }
+
+    UpdateCamera();
+
+    if (m_light == nullptr) {
+        RCFileClass file(nullptr, "Light.w3d");
+        W3DAssetManager::Get_Instance()->Load_3D_Assets(file);
+        m_light = W3DAssetManager::Get_Instance()->Create_Render_Obj("LIGHT");
+        m_lightInScene = false;
+    }
+
+    m_isInitialized = success;
+
+    if (success && m_timer == 0) {
+        timecaps_tag ptc;
+        ptc.wPeriodMin = 0;
+        ptc.wPeriodMax = 0;
+        timeGetDevCaps(&ptc, sizeof(ptc));
+        UINT time = ptc.wPeriodMin;
+
+        if (ptc.wPeriodMin <= 16) {
+            time = 16;
+        }
+
+        m_timer = timeSetEvent(time, time, TimerProc, (DWORD_PTR)m_hWnd, TIME_PERIODIC);
+    }
+
+    return success;
+}
+
+void CGraphicView::EnableRendering(bool enable)
+{
+    if (enable) {
+        m_renderingDisabled -= 1;
+    } else {
+        m_renderingDisabled += 1;
+    }
+}
+
+void ResetLODLevel(RenderObjClass *obj)
+{
+    if (obj != nullptr) {
+        for (int i = 0; i < obj->Get_Num_Sub_Objects(); i++) {
+            RenderObjClass *o = obj->Get_Sub_Object(i);
+
+            if (o != nullptr) {
+                ResetLODLevel(o);
+                Ref_Ptr_Release(o);
+            }
+        }
+
+        if (obj->Class_ID() == RenderObjClass::CLASSID_HLOD) {
+            obj->Set_LOD_Level(0);
+        }
+    }
+}
+
+void CGraphicView::Render(BOOL update, unsigned int time)
+{
+    static bool IsRendering = false;
+
+    if (!IsRendering) {

```

Comment:
Does this condition serve any purpose? It looks like it does nothing. Can it be removed along with `IsRendering`?

Reply:
IsRendering is static so its designed to make sure Render isn't run if its already running.